crates/rolldown/src/module_loader/module_loader.rs

```diff
@@ -158,119 +187,132 @@ impl ModuleLoader {
     assert_module_type: Option<ModuleType>,
     user_defined_entries: &[(Option<ArcStr>, ResolvedId)],
   ) -> ModuleIdx {
-    match self.visited.get(&resolved_id.id) {
-      Some(VisitState::Seen(idx)) => *idx,
-      Some(VisitState::Invalidate(_idx)) => unimplemented!("TODO: handle invalidation"),
+    let idx = match self.visited.get(&resolved_id.id) {
+      Some(VisitState::Seen(idx)) => return *idx,
+      Some(VisitState::Invalidate(idx)) => {
+        // Full scan mode the idx will never be invalidated right?
+        let idx = *idx;
+        self.intermediate_normal_modules.alloc_ecma_module_idx_sparse(idx);
+        self.visited.insert(resolved_id.id.clone(), VisitState::Seen(idx));
+        idx
+      }
+      None if !self.is_full_scan => {
+        // This means some new module has been added in partial scan mode
+        let len = self.visited.len();
+        let idx = self.intermediate_normal_modules.alloc_ecma_module_idx_sparse(len.into());
+        self.visited.insert(resolved_id.id.clone(), VisitState::Seen(idx));
+        idx
+      }
```

Comment:
A lot of diff came from I extract if else branch out of match block, It maybe more helpful if you use a semantic diff tool https://app.semanticdiff.com/gh/rolldown/rolldown/pull/3934/changes#crates/rolldown/src/module_loader/module_loader.rs?ignore_comments=false